test/infrastructure/docker/exp/internal/controllers/dockermachinepool_controller_phases.go

```diff
@@ -0,0 +1,356 @@
+// - Ensuring each Docker container has an associated DockerMachine by creating one if it doesn't already exist.
+// - Ensuring that deletion for Docker container happens by calling delete on the associated Machine so that the node is cordoned/drained and the infrastructure is cleaned up.
+// - Deleting DockerMachines referencing a container whose Kubernetes version or custom image no longer matches the spec.
+// - Deleting DockerMachines that correspond to a deleted/non-existent Docker container.
+// - Deleting DockerMachines when scaling down such that DockerMachines whose owner Machine has the clusterv1.DeleteMachineAnnotation is given priority.
+func (r *DockerMachinePoolReconciler) reconcileDockerMachines(ctx context.Context, cluster *clusterv1.Cluster, machinePool *expv1.MachinePool, dockerMachinePool *infraexpv1.DockerMachinePool) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	log.V(2).Info("Reconciling DockerMachines", "dockerMachinePool", dockerMachinePool.Name, "namespace", dockerMachinePool.Namespace)
+
+	dockerMachineList, err := getDockerMachines(ctx, r.Client, *cluster, *machinePool, *dockerMachinePool)
+	if err != nil {
+		return err
+	}
+
+	dockerMachineSet := make(map[string]infrav1.DockerMachine)
+	for _, dockerMachine := range dockerMachineList.Items {
+		dockerMachineSet[dockerMachine.Name] = dockerMachine
+	}
+
+	labelFilters := map[string]string{dockerMachinePoolLabel: dockerMachinePool.Name}
+
+	// List the Docker containers. This corresponds to a InfraMachinePool instance for providers.
+	externalMachines, err := docker.ListMachinesByCluster(ctx, cluster, labelFilters)
+	if err != nil {
+		return errors.Wrapf(err, "failed to list all machines in the cluster")
+	}
+
+	// Step 1:
+	// Create a DockerMachine for each Docker container so we surface the information to the user. Use the same name as the Docker container for the Docker Machine for ease of lookup.
+	// Providers should iterate through their infrastructure instances and ensure that each instance has a corresponding InfraMachine.
+	for _, machine := range externalMachines {
+		if _, ok := dockerMachineSet[machine.Name()]; !ok {
+			log.V(2).Info("Creating a new DockerMachine for Docker container", "container", machine.Name())
+			dockerMachine, err := r.createDockerMachine(ctx, machine.Name(), cluster, machinePool, dockerMachinePool)
+			if err != nil {
+				return errors.Wrap(err, "failed to create a new docker machine")
+			}
+
+			dockerMachineSet[dockerMachine.Name] = *dockerMachine
+		} else {
+			log.V(4).Info("DockerMachine already exists, nothing to do", "name", machine.Name())
+		}
+	}
+
+	// Step 2:
+	// Delete any DockerMachines that correspond to an outdated Docker container, meaning the K8s version or custom image of the container no longer matches the spec.
+	// If the InfraMachinePool spec changes, providers should delete InfraMachines (and owner Machines) for an instance that no longer matches the spec.
+	for _, dockerMachine := range dockerMachineSet {
```

Comment:
Would it be possible to break out the minimum changes from this PR required to fix https://github.com/kubernetes-sigs/cluster-api/issues/9655?

Reply:
I don't really believe we can really split up this PR as we weren't creating DockerMachines at all, instead we relied on [nodepool.go](https://github.com/kubernetes-sigs/cluster-api/blob/dec1667cb1c86bc6fbf464926b5e27886be73da3/test/infrastructure/docker/exp/internal/docker/nodepool.go) which was a black box that created/deleted the containers. It proved very awkward to refactor nodepool to support DockerMachines, so I ended up refactoring it into what we have now.

Reply:
Agree with @Jont828 that it's not splittable, what introduces proper cordon and drain of DockerMachinePools _is_ MachinePoolMachines as we're now reusing the Machine controller logic for cordon and drain